### DISCUSSION TITLE
ROWY Basic Webhook Parse javaScript : how to add a row populated by JSON Body ?

### DISCUSSION THREAD

**Discussion by Olmatro** [AntlerVC/firetable#795] [Support / Q&A] [answered]
ROWY Basic Webhook Parse javaScript : how to add a row populated by JSON Body ?

Hi,

Here is the Subject of my Question:
I am currently trying to receive a JSON Body (which I have previously tested in both webhook.site and hookdeck.com to diversify webhook's reception test applications) in ROWY Webhooks to:
- add a row to my ROWY Table ;
- populate the Text Fields (PK, PICTURE-MetaData, airtableRecID, airtablePictureURL) of this new row with the "Key": "Value" of the received JSON Body.

To do that I created and deployed a ROWY BASIC Webhook which I did not modify the PARSE javascript because I could not find examples corresponding to my useCase
or a TUTO to do it correctly.

But I only get the Add New Row to my ROWY Table.
Never the Text Fields (PK, PICTURE-MetaData, airtableRecID, airtablePictureURL) of this new row are populated with the "Key": "Value" of the received JSON Body.

I don't have access to error messages from ROWY after receiving this Webhook Body and I can't read any console.log that I would try to add to the PARSE Code since this code is executed on the Server side so Chrome Dev Tools Javascript Console doesn't help me in this case.

You will find the Documentation of my case below:

1) my posted webHook received by webhook.site:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00001](https://user-images.githubusercontent.com/13352905/189777041-3584a5d7-3d98-499f-8f5f-ca3004ae40b6.png)

2) my posted webHook received by hookdeck:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00005](https://user-images.githubusercontent.com/13352905/189777196-eacf04c4-64ce-4831-9431-60fbc00ce6c8.png)
2-1) hookdeck event data:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00007](https://user-images.githubusercontent.com/13352905/189777283-eab6a32c-65fd-4caa-a4a5-5eae59244ed9.png)
2-2) hookdeck received JSON Body: 
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00008](https://user-images.githubusercontent.com/13352905/189777451-24cbaa7b-caa9-479b-be55-b21484645d1a.png)

-> According to the audit of my webhook performed by 1) and 2), my webhook seems compatible with a good reception by an endpoint. 

3) Here is the Script that is posting from Airtable Scripting Extension to webhook:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00026](https://user-images.githubusercontent.com/13352905/189780275-157dcead-2dbb-450e-bea1-4d7380b8ad4c.png)
3-1) What this script is actually posting:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00025](https://user-images.githubusercontent.com/13352905/189780387-78e36b1c-9c07-485c-8fe9-194f60486de7.png)
3-2) JSON Body detail of 3-1:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00024](https://user-images.githubusercontent.com/13352905/189780561-64cd4c24-5212-4ae8-8415-521c948b50dd.png)

4) Here is my ROWY Table, manually populated (4 ROWs) at this time:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00013](https://user-images.githubusercontent.com/13352905/189780811-c1fa018d-7dba-4fa8-8109-b7a5ad4e5433.png)

![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00014](https://user-images.githubusercontent.com/13352905/189780876-b9c03bb9-fb0f-4e7e-9d4f-47a8d8133f9a.png)

5) Let's go to ROWY WEBHOOK menu:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00015](https://user-images.githubusercontent.com/13352905/189781378-7580e5b6-7a3f-4192-9091-683fa290c6e2.png)

![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00016](https://user-images.githubusercontent.com/13352905/189781427-7a865f31-1c2a-48e3-a888-b38ef3f91791.png)

6) Let's Edit this Webhook:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00017](https://user-images.githubusercontent.com/13352905/189781581-fd75e11d-4ac8-43b5-8fa5-b98db150dc36.png)

7) to take a look at BASIC WEBHOOK PARSE Script:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00018](https://user-images.githubusercontent.com/13352905/189781768-592ad9a8-3ca9-4096-9b50-d19aa5c08218.png)

8) Please note that this Webhook was correctly UPDATED and SAVED and DEPLOYED.

9) Please take a look at ROWY TABLE just BEFORE POSTing to Webhook URL:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00019](https://user-images.githubusercontent.com/13352905/189782084-ed0ecc57-c185-451c-abde-c955a134949e.png)

10) Please take a look at ROWY TABLE just AFTER POSTing to Webhook URL:
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00021](https://user-images.githubusercontent.com/13352905/189782188-1bf00e43-70da-4d2d-b638-9bd9b03deb5c.png)

Would you notice that a NEW ROW appeared on top of TABLE "created by Rowy Hooks" ?

11) Unfortunately, the Text Fields PK, PICTURE-MetaData, airtableRecID, airtablePictureURL corresponding to "Key": "Value" in the JSON have NOT been populated : they remain empty !
Only _createdBy has been populated by the Script designed to do that.
![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00022](https://user-images.githubusercontent.com/13352905/189782935-9983ef3c-7ff5-4166-b611-f1bac7c5ac89.png)

![220912_18h10-hello-at-rowy-io-webHooks-rowy-as-airtable-cdn-ui-00023](https://user-images.githubusercontent.com/13352905/189782966-160aaa30-4af4-4e70-8bbb-1558d9061118.png)

12) At this point, I'm stuck because I can't find a more appropriate javascript to get :
- the addition of a Line at the Webhook Reception by ROWY ;
AND
- Getting the "Key": "Value" of the JSON to populate the PK, PICTURE-MetaData, airtableRecID, airtablePictureURL Text Fields of the new row.

You may note that since I have complete freedom to modify the source code of the POST, I can also create a more appropriate JSON Body to make it work better.

I have read everything I could find about ROWY Webhook but I couldn't find enough examples of code simpler than Typeform to inspire me.

Please help me to improve both the JSON BODY at the source and the PARSE of ROWY Webhook 
and thank you too for pointing me to a Console that would show me errors where they would occur on the ROWY side.
Regards,
oLpY

**Comment by organicnz** [comment, marked as answer]
@Olmatro hi mate. Could you pls share how you've populated something like the `dictionaryApi` column by any chance? :)

I'm not getting any JSON data after the evaluation :) 

https://www.loom.com/i/802dccd25e6344dcaae22ff771a2a664

https://www.loom.com/i/200543cd0f4943469111b83858ea538e

**Reply by Olmatro** [reply]
Hi,
I suppose you would take a look at this DERIVATIVE Code ?

```
const derivative:Derivative = async ({row,ref,db,storage,auth})=>{
    // Write your derivative code here
    // for example:
    // const sum = row.a + row.b;
    const pictureUrl = row.picture[0].downloadURL;
    // return sum;
    return pictureUrl;
    // checkout the documentation for more info: https://docs.rowy.io/field-types/derivative
  }
```

related to my 
![Screen Shot 2022-09-13 at 16 50 03](https://user-images.githubusercontent.com/13352905/189932301-601d922d-d52d-44cc-afcc-d6ca280ffcbe.png)
**PICTURE-URL** field ?

How did I reach that property ?
By taking a look at this, then a small amount of try-error in DERIVATIVE javascript :
![Screen Shot 2022-09-13 at 16 55 39](https://user-images.githubusercontent.com/13352905/189937537-c26a0cc6-8162-4f93-9dac-70e33d1f29e8.png)

![Screen Shot 2022-09-13 at 16 56 08](https://user-images.githubusercontent.com/13352905/189940057-6633ccaa-e997-406b-a545-f0052b7ef8fc.png)

I mostly like to explore, take a (long) look to objects, data, then trying small code.
Despite it, Webhooks PARSE is still winning against that way of going further.

Cheers,
oLπ

**Reply by organicnz** [reply]
Hey @Olmatro, thanks for getting back, mate. I should learn about Webhooks parsing then hehe :)

Sorry that I haven't made it quite clear. I'd like to get the JSON data and then data extraction would not be a problem. I seem to be doing something wrong as it doesn't return anything lol

```
const derivative:Derivative = async ({row,ref,db,storage,auth})=>{
    const res = await fetch(`https://api.airtable.com/v0/appxQQHeIXAeUI0wI/Fridges?api_key=xxxxxxxxxxxxx`);
    const json = await res.json();
    return json;
  }
```

![8dd2674c6ae949d386d8ef5a75d0131b](https://user-images.githubusercontent.com/4765090/190217692-9e5c2921-0581-4b9e-918f-891f8621f824.png)

**Reply by Olmatro** [reply]
Hi @organicnz ,

You are right : I didn't understand you well! ;)
What is more complicated in your case, is that you can't check if the JSON payload you just fetched is empty, like I could do with webhooks.site or with another similar tool,
or if there is an error code in return of your FETCH, or if it is ROWY which does not transmit your JSON payload to your Field !
I would be interested to try the same thing using the same code from an Airtable Scripting Extension
because await Fetch to Airtable API can be run from a Scripting Extension too.

If your JSON payload received through the FETCH was not empty, you are highlighting a problem of the same type that I had encountered during my adventure with the webHooks:
When ROWY doesn't return anything, WHERE to look / find some kind of Log and Console to analyze / debug what went wrong?
I suspect this exists in the Google Cloud Platform environment, but I'll have to search a bit in GCP.
I think I'll have some free time to investigate this a bit next Saturday and/or Sunday.

oLπ

**Reply by organicnz** [reply]
> Hi @organicnz ,
> 
> You are right : I didn't understand you well! ;) What is more complicated in your case, is that you can't check if the JSON payload you just fetched is empty, like I could do with webhooks.site or with another similar tool, or if there is an error code in return of your FETCH, or if it is ROWY which does not transmit your JSON payload to your Field ! I would be interested to try the same thing using the same code from an Airtable Scripting Extension because await Fetch to Airtable API can be run from a Scripting Extension too.
> 
> If your JSON payload received through the FETCH was not empty, you are highlighting a problem of the same type that I had encountered during my adventure with the webHooks: When ROWY doesn't return anything, WHERE to look / find some kind of Log and Console to analyze / debug what went wrong? I suspect this exists in the Google Cloud Platform environment, but I'll have to search a bit in GCP. I think I'll have some free time to investigate this a bit next Saturday and/or Sunday.
> 
> oLπ

Thanks for the detailed explanation, mate. In fact, I was able to fix it today by importing the data using the newly added Rowy function. The data has now been filled in :)

<img width="1440" alt="Screen Shot 2022-10-05 at 11 49 54 PM" src="https://user-images.githubusercontent.com/4765090/194161072-2e014f24-971f-4a12-a03f-6f8b830878d2.png">
